File: Euler_114_Counting_Block_Combinations_I/go/euler-114-counting-block-combinations-I.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func main() {
	size, _ := strconv.ParseInt(os.Args[1], 10, 64)
	fmt.Printf("Answer: %d\n", CheckBlocks(int(size)))
}

// possibleArr[i] - an array to store number of possible combinations of tile of length i
// i.e. possibleArr[1] = tile with length of 1 and this has only 1 possible combination (all gray)
// Iterate through all of sizes up to needed size to find the amount of combinations.
func CheckBlocks(size int) int64 {
	possibleArr := [51]int64{1, 1, 1}
	for i := 3; i <= size; i++ {
		sum := possibleArr[i-1] + 1
		for j := 3; j < i; j++ {
			sum = sum + possibleArr[i-j-1]
		}
		possibleArr[i] = sum
	}
	return possibleArr[size]
}

File: Euler_114_Counting_Block_Combinations_I/java/euler114CountingBlockCombinationsI.java
public class euler114CountingBlockCombinationsI {
    public static void main(String[] args) {
        int size = Integer.parseInt(args[0]);
        System.out.println("Answer: " + checkBlocks(size));
        
    }

    static long checkBlocks(int size) {
        long[] possibleArr = new long[size + 1];
        possibleArr[0] = 1;
        possibleArr[1] = 1;
        possibleArr[2] = 1;
        for(int i=3; i<=size; i++) {
            long sum = possibleArr[i-1] + 1;
            for(int j=3; j<i; j++) {
                sum = sum + possibleArr[i - j - 1];
            }
            possibleArr[i] = sum;
        }
        return possibleArr[size];
    }
}